Runtime/Tracking/Collision/Active/ActiveCollisionsContainer.cs

```diff
@@ -196,15 +224,71 @@ protected virtual CollisionNotifier.EventData CloneEventData(CollisionNotifier.E
         }
 
         /// <summary>
-        /// Determines if the current collision is valid.
+        /// Determines if the current containing <see cref="Transform"/> in the collision is valid.
         /// </summary>
-        /// <param name="collisionData">The collision to check.</param>
+        /// <param name="containingTransform">The <see cref="Transform"/> to check.</param>
         /// <returns>The validity result of the collision.</returns>
-        protected virtual bool IsValidCollision(CollisionNotifier.EventData collisionData)
+        protected virtual bool IsValidCollision(Transform containingTransform)
         {
-            Transform containingTransform = collisionData.ColliderData.GetContainingTransform();
             return containingTransform != null && CollisionValidity.Accepts(containingTransform.gameObject);
         }
 
+        /// <summary>
+        /// Adds the given data to the <see cref="Elements"/> collection if it is not already present.
+        /// </summary>
+        /// <param name="data">The data to add.</param>
+        protected virtual void AddUniqueElement(CollisionNotifier.EventData data)
+        {
+            if (!Elements.Contains(data))
+            {
+                Elements.Add(data);
+            }
+        }
+
+        /// <summary>
+        /// Determines if the containing <see cref="Transform"/> is already being collided with by a different nested <see cref="Collider"/>.
+        /// </summary>
+        /// <param name="collisionData">The collision data containing the <see cref="Collider"/> to check.</param>
+        /// <param name="currentCollisionContainingTransform">The containing <see cref="Transform"/> for the current collider.</param>
+        /// <returns>Whether there are any existing collisions occuring with the containing <see cref="Transform"/>.</returns>
+        protected virtual bool AreExistingCollisions(CollisionNotifier.EventData collisionData, Transform currentCollisionContainingTransform)

```

Comment:
This method does way more than doing a simple lookup/check as the name implies. It feels worth to rename it to explicitly describe the state mutation it's doing and keep the return value documented via the XML docs. Suggestion: `AddCollisionData`

Reply:
It's not actually adding collision data persay.

It's checking if there are any existing collisions for the given ContainingTransfomr and if there are no collisions at all for the containing transform then it returns false.

If there are collisions for the containing transform that match directly for the current collision then it returns true.

The rest of it is just maintaining the `Elements` array so the latest collision for the containing transform is stored.

Perhaps instead the `AreExistingCollisions` can stay as is, but the mutation of `Elements` can be moved into its own method (e.g. UpdateContainerCollisions)

Then the call could be something like

```
if (AreExistingCollisions(collisionData, currentCollisionContainingTransform)) {
 UpdateContainerCollisions(params)
} 
else {
...exiting below code
}
```

Reply:
👍

Reply:
lol I can't actually do what I suggested because `true` is returned twice but different logic happens so separating it out isn't going to help

I need the logic to be:

* Is there a collision for the containing transform already? no then return false
* There is a containing transform collision, but is there a collision for the current collision data? yes, then return true
* If there is a containing transform collision but a new nested collision then update the `Elements` with this new collision and remove any old collisions, then return true

If i split out that 3rd point then the logic will all be wrong.

I could still move the Remove/Add logic to a sub method to make it easier to read, but that means then the `AreExistingCollisions` method is still internally mutating that `Elements` collection so we're back to "the name doesn't really say what's fully going on"